### ISSUE TITLE
java.lang.RuntimeException: com.intellij.serviceContainer.AlreadyDisposedException: Cannot create com.intellij.openapi.roots.impl.libraries.ProjectLibraryTable because container is already disposed (container=Project(name=dealer-portal, containerState=DISPOSE_COMPLETED, componentStore=/Users/jgutierrez/ElixirPhoenix/dealer-portal) (disposed)) at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:118)

### ISSUE THREAD

**Issue by Chucheen** [KronicDeth/intellij-elixir#3389]
# System

Plugin Version: 13.2.1
Application: IntelliJ IDEA Ultimate Edition (2022.2.3)
Operating System: Mac OS X (12.4)

# Event

### Stacktrace

<details>

<summary>
java.lang.RuntimeException: com.intellij.serviceContainer.AlreadyDisposedException: Cannot create com.intellij.openapi.roots.impl.libraries.ProjectLibraryTable because container is already disposed (container=Project(name=dealer-portal, containerState=DISPOSE_COMPLETED, componentStore=/Users/jgutierrez/ElixirPhoenix/dealer-portal) (disposed))
</summary>

```
java.lang.RuntimeException: com.intellij.serviceContainer.AlreadyDisposedException: Cannot create com.intellij.openapi.roots.impl.libraries.ProjectLibraryTable because container is already disposed (container=Project(name=dealer-portal, containerState=DISPOSE_COMPLETED, componentStore=/Users/jgutierrez/ElixirPhoenix/dealer-portal) (disposed))
	at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:118)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:505)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:525)
	at org.elixir_lang.mix.Watcher.syncLibraries(Watcher.kt:70)
	at org.elixir_lang.mix.Watcher.syncLibraries(Watcher.kt:65)
	at org.elixir_lang.DepsWatcher.syncLibraries(DepsWatcher.kt:198)
	at org.elixir_lang.DepsWatcher.syncLibraries(DepsWatcher.kt:181)
	at org.elixir_lang.DepsWatcher.syncLibraries(DepsWatcher.kt:176)
	at org.elixir_lang.DepsWatcher$syncLibraries$2$2$1.run(DepsWatcher.kt:117)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$5(CoreProgressManager.java:493)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:188)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:175)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.intellij.serviceContainer.AlreadyDisposedException: Cannot create com.intellij.openapi.roots.impl.libraries.ProjectLibraryTable because container is already disposed (container=Project(name=dealer-portal, containerState=DISPOSE_COMPLETED, componentStore=/Users/jgutierrez/ElixirPhoenix/dealer-portal) (disposed))
	at com.intellij.serviceContainer.ContainerUtilKt.throwAlreadyDisposedError(containerUtil.kt:34)
	at com.intellij.serviceContainer.ComponentManagerImpl.doGetService(ComponentManagerImpl.kt:631)
	at com.intellij.serviceContainer.ComponentManagerImpl.getService(ComponentManagerImpl.kt:583)
	at com.intellij.openapi.client.ClientAwareComponentManager.getFromSelfOrCurrentSession(ClientAwareComponentManager.kt:37)
	at com.intellij.openapi.client.ClientAwareComponentManager.getService(ClientAwareComponentManager.kt:22)
	at com.intellij.openapi.roots.impl.libraries.LibraryTablesRegistrarImpl.getLibraryTable(LibraryTablesRegistrarImpl.java:41)
	at org.elixir_lang.mix.Watcher.syncLibraries$lambda-9$lambda-8(Watcher.kt:72)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1023)
	at org.elixir_lang.mix.Watcher.syncLibraries$lambda-9(Watcher.kt:71)
```

</details>

**Comment by KronicDeth** [comment]
This version of the plugin is no longer supported.

**Comment by sh41** [comment]
This is the same crash as #1850, and it has since been fixed. `Watcher.syncLibraries` used
`invokeAndWait` to queue a write action from a background task without checking the project was still
alive when it ran, so closing a project mid-sync reached `LibraryTablesRegistrar.getLibraryTable` on a
container already disposed.

`f4e399504` replaced `Watcher` and `DepsWatcher` with `MixDepsSyncService`, which checks
`project.isDisposed` immediately before that call and no longer blocks on `invokeAndWait`. First
released in 24.0.0 — well after the 13.2.1 you reported this against.

Leaving this closed and recording the fix on #1850, the canonical report for this crash.

This issue was looked at as part of an AI-assisted review of the backlog, and a human checked this
reply before it was posted. If any of it is wrong, please say so here.
